jit/jit.cpp

```diff
@@ -0,0 +1,467 @@
+#include <llvm/IR/Type.h>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/ExecutionEngine/JITSymbol.h>
+#include <llvm/ExecutionEngine/SectionMemoryManager.h>
+#include <llvm/ExecutionEngine/JITEventListener.h>
+#include <llvm/MC/SubtargetFeature.h>
+#include <llvm/MC/TargetRegistry.h>
+#include <llvm/Support/DynamicLibrary.h>
+#include <llvm/Support/Host.h>
+#include <llvm/Support/TargetSelect.h>
+#include <llvm/Transforms/IPO/PassManagerBuilder.h>
+#include <llvm/Support/SmallVectorMemoryBuffer.h>
+#include <llvm/Passes/PassBuilder.h>
+
+#include <llvm/ExecutionEngine/RTDyldMemoryManager.h>
+#include <llvm/ADT/None.h>
+#include <llvm/ExecutionEngine/MCJIT.h>
+
+#include <boost/noncopyable.hpp>
+namespace msc_jit
+{
+// Arena Memory pool for JIT
+class Arena : private boost::noncopyable
+{
+ public:
+  Arena() : page_size(8)
+  {
+  }
+  ~Arena()
+  {
+    protect(PROT_READ | PROT_WRITE);
+
+    for (auto& block : blocks)
+    {
+      free(block.base());
+    }
+  }
+  /*
+   * param: size: the size of the memory to be allocated
+   * return: the address of the allocated memory
+   * */
+  char* allocateAligned(size_t size, size_t alignment)
+  {
+    for (size_t i = 0; i < blocks.size(); ++i)
+    {
+      char* result = allocateFromBlockWithIndex(size, alignment, i);
+      if (result)
+      {
+        return result;
+      }
+    }
+    allocateNewBlock(size);
+    size_t allocated_index = blocks.size() - 1;
+    char* result = allocateFromBlockWithIndex(size, alignment, allocated_index);
+    assert(result);
+    return result;
+  }
+
+  void protect(int flag)
+  {
+#if defined(__NetBSD__) && defined(PROT_MPROTECT)
+    flag |= PROT_MPROTECT(PROT_READ | PROT_WRITE | PROT_EXEC);
+#endif
+
+    bool invalidate_cache = (flag & PROT_EXEC);
+
+    for (const auto& block : blocks)
+    {
+#if defined(__arm__) || defined(__aarch64__)
+      /// Certain ARM implementations treat icache clear instruction as a memory read,
+      /// and CPU segfaults on trying to clear cache on !PROT_READ page.
+      /// Therefore we need to temporarily add PROT_READ for the sake of flushing the instruction caches.
+      if (invalidate_cache && !(flag & PROT_READ))
+      {
+        int res = mprotect(block.base(), block.blockSize(), flag | PROT_READ);
+        if (res != 0)
+          throw std::runtime_error("Cannot mprotect memory region");
+
+        llvm::sys::Memory::InvalidateInstructionCache(block.base(), block.blockSize());
+        invalidate_cache = false;
+      }
+#endif
+      int res = mprotect(block.base(), block.blockSize(), flag);
+      if (res != 0)
+        throw std::runtime_error("Cannot mprotect memory region");
+
+      if (invalidate_cache)
+        llvm::sys::Memory::InvalidateInstructionCache(block.base(), block.blockSize());
+    }
+  }
+
+  inline size_t getAllocatedSize() const
+  {
+    return allocated_size;
+  }
+
+  inline size_t getPageSize() const
+  {
+    return page_size;
+  }
+
+ private:
+  size_t page_size = 0;
```

Comment:
I do not see any modification of this `page_size` in the code. I think that it could be declared simply as
```c++
  const size_t page_size = 8;
```
or even `static const` or `static constexpr` to save some memory, and then we could avoid explicitly declaring `Arena()`. I think that it is rather pointless to declare `const` data members `private` and to define public accessors for them.

What is the unit here? Bytes? Kilobytes? Some comments would be helpful.

Typically, `sysconf(PAGESIZE)` should return 4096 or a larger value.